test/integration/consul-container/test/gateways/terminating_gateway_test.go

```diff
@@ -0,0 +1,201 @@
+		_, _, err := consulClient.Namespaces().Create(&api.Namespace{
+			Name:      namespace,
+			Partition: part,
+		}, nil)
+		require.NoError(t, err)
+	}
+
+	t.Logf("registering the external service")
+	_, err := consulClient.Catalog().Register(&api.CatalogRegistration{
+		Node:     "legacy_node",
+		Address:  address,
+		NodeMeta: map[string]string{"external-node": "true", "external-probe": "true"},
+		Service:  service,
+	}, nil)
+	require.NoError(t, err)
+}
+
+// createTerminatingGatewayConfigEntry creates and sets a config entry that binds the
+// services to the gateway.
+func createTerminatingGatewayConfigEntry(t *testing.T, consulClient *api.Client, gwNamespace, serviceNamespace string, serviceNames ...string) {
+	t.Helper()
+
+	t.Logf("creating terminating gateway config entry")
+
+	if serviceNamespace != "" {
+		t.Logf("creating the %s namespace in Consul", serviceNamespace)
+		_, _, err := consulClient.Namespaces().Create(&api.Namespace{
+			Name: serviceNamespace,
+		}, nil)
+		require.NoError(t, err)
+	}
+
+	var gatewayServices []api.LinkedService
+	for _, serviceName := range serviceNames {
+		linkedService := api.LinkedService{Name: serviceName, Namespace: serviceNamespace}
+		gatewayServices = append(gatewayServices, linkedService)
+	}
+
+	configEntry := &api.TerminatingGatewayConfigEntry{
+		Kind:      api.TerminatingGateway,
+		Name:      "terminating-gateway",
+		Namespace: gwNamespace,
+		Services:  gatewayServices,
+	}
+
+	created, _, err := consulClient.ConfigEntries().Set(configEntry, nil)
+	require.NoError(t, err)
+	require.True(t, created, "failed to create terminating gateway config entry")
+}
+
+// assertHTTPRequestToServiceAddress checks the result of a request from the
+// given `client` container to the given `server` container. If expSuccess is
+// true, this checks for a successful request and otherwise it checks for the
+// error we expect when traffic is rejected by mTLS.
+//
+// This assumes the destination service is running Fortio. It makes the request
+// to `<serverIP>:8080/debug?env=dump` and checks for `FORTIO_NAME=<expServiceName>`
+// in the response.
+func assertHTTPRequestToServiceAddress(t *testing.T, client *libservice.ConnectContainer, serviceName string, port int, expSuccess bool) {
+	upstreamURL := fmt.Sprintf("http://localhost:%d/debug?env=dump", port)
+	retry.RunWith(requestRetryTimer, t, func(r *retry.R) {
+		out, err := client.Exec(context.Background(), []string{"curl", "-s", upstreamURL})
+		t.Logf("curl request to upstream service address: url=%s\nerr = %v\nout = %s", upstreamURL, err, out)
+
+		if expSuccess {
+			require.NoError(r, err)
+			require.Contains(r, out, fmt.Sprintf("FORTIO_NAME=%s", serviceName))
+			t.Logf("successfuly messaged %s", serviceName)
+		} else {
+			require.Error(r, err)
+			require.Contains(r, err.Error(), "exit code 52")
+		}
+	})
+}
+
+func createNamespace(t *testing.T, consulClient *api.Client, namespace string) {
```

Comment:
I think actually this should be in an enterprise only file, you might want to refactor this?

Reply:
Sure!

Reply:
I've moved it out and into the other ENT PR.